File: frontend/src/types/testplan.ts
// 测试计划相关类型

export type PlanStatus = 'draft' | 'ready' | 'running' | 'completed' | 'cancelled'
export type ExecutionStatus = 'pending' | 'running' | 'passed' | 'failed' | 'skipped' | 'error'
export type TriggerType = 'manual' | 'cron' | 'zentao_webhook' | 'api'

export interface TestPlan {
  id: number
  projectId: number
  name: string
  description: string
  caseIds: number[]
  caseCount: number
  status: PlanStatus
  cronExpr?: string      // 定时执行 cron 表达式
  maxRetries: number     // 失败重试次数
  timeout: number        // 整体超时（秒）
  createdBy: number
  creatorName?: string
  createdAt: string
  updatedAt: string
  lastRunAt?: string
}

export interface Execution {
  id: number
  planId: number
  planName?: string
  caseId: number
  caseTitle?: string
  status: ExecutionStatus
  triggerType: TriggerType
  startTime?: string
  endTime?: string
  duration?: number      // 执行耗时（ms）
  errorMessage?: string
  screenshots: string[]  // 截图路径
  domSnapshot?: string   // 失败时 DOM 快照
  log: string            // 执行日志
  steps: ExecutionStep[]
  zentaoBugId?: number   // 上报禅道的 Bug ID
  retryCount: number
  executedBy: string     // manual / cron / webhook
}

export interface ExecutionStep {
  seq: number
  action: string
  status: 'passed' | 'failed' | 'skipped'
  screenshot?: string
  errorMessage?: string
  duration?: number
  timestamp: string
}

File: backend/app/schemas/testplan.py
"""测试计划相关 Schema"""

from pydantic import BaseModel
from datetime import datetime


class TestPlanCreate(BaseModel):
    project_id: int
    name: str
    description: str | None = None
    case_ids: list[int] = []
    cron_expr: str | None = None
    max_retries: int = 2
    timeout: int = 600


class TestPlanUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    case_ids: list[int] | None = None
    status: str | None = None
    cron_expr: str | None = None
    max_retries: int | None = None
    timeout: int | None = None


class TestPlanResponse(BaseModel):
    id: int
    project_id: int
    name: str
    description: str | None = None
    case_ids: list
    case_count: int = 0
    status: str
    cron_expr: str | None = None
    max_retries: int
    timeout: int
    created_by: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
